src/videoframeprocessor/CanvasVideoFrameBuffer.ts

```diff
@@ -0,0 +1,32 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import VideoFrameBuffer from './VideoFrameBuffer';
+
+/**
+ * [[CanvasVideoFrameBuffer]] implements [[VideoFrameBuffer]]. It internally holds an `HTMLCanvasElement`.
+ */
+export default class CanvasVideoFrameBuffer implements VideoFrameBuffer {
+  private destroyed: boolean = false;
+  framerate: number;
+  width: number;
+  height: number;
+
+  constructor(private canvas: HTMLCanvasElement) {}
+
+  destroy(): void {
+    this.canvas = null;
+    this.destroyed = true;
+  }
+
+  asCanvasImageSource(): Promise<CanvasImageSource> {
+    if (this.destroyed) {
+      throw Error('canvas buffer is destroyed');
```

Comment:
Should this be Promise.reject here?

Reply:
is it one preferred over the other ?

Reply:
I guess not. I think we use both way in our code right now.

Reply:
If you throw inside a function that returns a `Promise`, you throw.

If you throw inside an `async` function that returns a `Promise`, you reject.

There is a difference; if you write

```typescript
function doSomething() {
  const p = foo.asCanvasImageSource();
  p.then(…);
  p.catch(…);
}
```

you will have `doSomething` throw if `aCIS` isn't marked `async`, and not throw if it is; the `catch` handler will run instead.

Reply:
You should mark `asCanvasImageSource` as `async`.

Reply:
okay, changed this `asCanvasImageSource` to `async` and change the `throw` to `Promise.reject`

Reply:
Just to be clear:

```
async function foo(): Promise<void> {
  throw new Error('x');
}
```

will return a promise. The `async` annotation affects error handling.